## Table of Contents

1. <a href="#item1">Web-scrape and Explore Dataset</a>

2. <a href="#item2">Fetch Latitude and Longitude of each Neighborhood</a>

3. <a href="#item3">Explore and Cluster the Neighborhoods in Toronto</a>

4. <a href="#item4">Analyze Each Neighborhood</a>

5. <a href="#item5">Cluster Neighborhoods</a>

6. <a href="#item6">Examine Clusters</a>   

Download all the dependencies.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import requests # library to handle requests
from bs4 import BeautifulSoup  # library to handle web scraping

import os
import wget

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import folium # map rendering library

import matplotlib.cm as cm # Matplotlib and associated plotting modules
import matplotlib.colors as colors # Matplotlib and associated plotting modules

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

from sklearn.cluster import KMeans # import k-means from clustering stage

<a id='item1'></a>

## 1. Web-scrape and Explore Dataset

Fetch the content of the url - https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [2]:
URL = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
r = requests.get(URL) 

Use BeautifulSoup library to read the content of the url.

In [3]:
soup = BeautifulSoup(r.content, 'html5lib')
# print(soup.prettify()) # visual representation of the parse tree

Inspect the HTML script above, the table of the postal codes is under class **Wikitable Sortable**.

In [4]:
table = soup.find('table', attrs = {'class':'wikitable sortable'}) 
# print(table.prettify())

Loop through the table to fetch column names and rows in a list

In [5]:
# Store header in column_names
# column_names = [header.text.strip() for header in table.findAll('th')]
column_names = ['PostalCode', 'Borough', 'Neighborhood']
column_names

['PostalCode', 'Borough', 'Neighborhood']

In [6]:
rows_list = []

# Loop through table, fetch each of the 3 columns of a row
for row in table.findAll('tr'):
    row_data = row.find_all('td')
    if len(row_data) == 3:
        rows_list.append((row_data[0].text.strip(), row_data[1].text.strip(), row_data[2].text.strip()))

rows_list[:5]

[('M1A', 'Not assigned', 'Not assigned'),
 ('M2A', 'Not assigned', 'Not assigned'),
 ('M3A', 'North York', 'Parkwoods'),
 ('M4A', 'North York', 'Victoria Village'),
 ('M5A', 'Downtown Toronto', 'Harbourfront')]

In [7]:
# instantiate the dataframe with column_names & rows_list
toronto_data = pd.DataFrame(data = rows_list, columns = column_names)
toronto_data.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

> * Process the cells that have an assigned borough. Ignore cells with a borough that is **Not assigned**.

In [8]:
toronto_data.shape

(288, 3)

In [9]:
print('The dataframe has {} unique boroughs and {} unique neighborhoods.'.format(
        len(toronto_data['Borough'].unique()),
        len(toronto_data['Neighborhood'].unique())
    )
)

The dataframe has 12 unique boroughs and 209 unique neighborhoods.


In [10]:
print(toronto_data.groupby('Borough').count().reset_index())

             Borough  PostalCode  Neighborhood
0    Central Toronto          17            17
1   Downtown Toronto          37            37
2       East Toronto           7             7
3          East York           6             6
4          Etobicoke          45            45
5        Mississauga           1             1
6         North York          38            38
7       Not assigned          77            77
8       Queen's Park           1             1
9        Scarborough          37            37
10      West Toronto          13            13
11              York           9             9


There are **77 boroughs** that are **Not assigned**.

In [11]:
toronto_data = toronto_data[toronto_data['Borough'] != 'Not assigned'].reset_index(drop=True)
toronto_data.head(10)

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights
5        M6A        North York    Lawrence Manor
6        M7A      Queen's Park      Not assigned
7        M9A         Etobicoke  Islington Avenue
8        M1B       Scarborough             Rouge
9        M1B       Scarborough           Malvern

In [12]:
# CONFIRMS 77 boroughs that are Not assigned ignored
toronto_data.shape

(211, 3)

> * If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [13]:
toronto_data[toronto_data['Neighborhood'] == 'Not assigned']

PostalCode       Borough  Neighborhood
6        M7A  Queen's Park  Not assigned

In [14]:
toronto_data.loc[toronto_data['Neighborhood'] == 'Not assigned', 'Neighborhood'] = toronto_data['Borough']
toronto_data.head(10)

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights
5        M6A        North York    Lawrence Manor
6        M7A      Queen's Park      Queen's Park
7        M9A         Etobicoke  Islington Avenue
8        M1B       Scarborough             Rouge
9        M1B       Scarborough           Malvern

In [15]:
# CONFIRMS Not assigned neighborhood replaced with borough
toronto_data[toronto_data['Neighborhood'] == 'Not assigned']

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

In [16]:
toronto_data['Neighborhood'].count()

211

In [17]:
len(toronto_data['Neighborhood'].unique())

209

There are 209 unique Neighborhood but total Neighborhood count is 211.<br>
Explore 'Neighborhood' that occurs more than once.

In [18]:
toronto_data.groupby('Neighborhood').filter(lambda x: len(x) > 1)

PostalCode           Borough    Neighborhood
27         M5C  Downtown Toronto  St. James Town
115        M6N              York       Runnymede
145        M6S      West Toronto       Runnymede
190        M4X  Downtown Toronto  St. James Town

* More than one neighborhood can exist in one postal code area. <br>
* These rows will be combined into one row with the neighborhoods separated with a comma.

In [19]:
toronto_data['PostalCode'].count()

211

In [20]:
len(toronto_data['PostalCode'].unique())

103

In [21]:
toronto_data = toronto_data.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(', '.join).reset_index()

In [22]:
toronto_data.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [23]:
# CONFIRMS rows combined to one 'Postcode'
toronto_data.shape

(103, 3)

<a id='item2'></a>

## 2. Fetch Latitude and Longitude of each Neighborhood

Download the latitude & longitude data and save as **Geospatial_Coordinates.csv**

In [24]:
if not os.path.exists('Geospatial_Coordinates.csv'):
    wget.download('http://cocl.us/Geospatial_data')

In [25]:
# read csv into pandas dataframe
geospatial_coordinates = pd.read_csv('Geospatial_Coordinates.csv')
geospatial_coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [26]:
toronto_data = toronto_data.join(geospatial_coordinates.set_index('Postal Code'), on=['PostalCode'])
toronto_data.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [27]:
toronto_data[toronto_data['PostalCode'] == 'M5G']

PostalCode           Borough        Neighborhood   Latitude  Longitude
57        M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383

<a id='item3'></a>

## 3.Explore and Cluster the Neighborhoods in Toronto

#### Use geopy library to get the latitude and longitude values of Toronto, Ontario.

In [28]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Ontario are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, Ontario are 43.653963, -79.387207.


#### Create a map of Toronto, Ontario with neighborhoods superimposed on top.

In [29]:
# create map of Toronto, Ontario using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version

In [30]:
CLIENT_ID = 'XOKJYQNJOBDGS1WJQLVWK0KY2MX2LSBS2RQHYZ1IERYGZQM2' # your Foursquare ID
CLIENT_SECRET = 'A20AUTOPOO4M4NLEVPQ4L5LLCSTC4EQL2R5AOUQWQPSN0IJO' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: XOKJYQNJOBDGS1WJQLVWK0KY2MX2LSBS2RQHYZ1IERYGZQM2
CLIENT_SECRET:A20AUTOPOO4M4NLEVPQ4L5LLCSTC4EQL2R5AOUQWQPSN0IJO


#### Let's explore the first neighborhood in the dataframe.

Get the neighborhood's name.

In [31]:
toronto_data.loc[0, 'Neighborhood']

'Rouge, Malvern'

Get the neighborhood's latitude and longitude values.

In [32]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rouge, Malvern are 43.806686299999996, -79.19435340000001.


#### Now, let's get the top 100 venues that are in Rouge, Malvern within a radius of 500 meters.

First, let's create the GET request URL.

In [33]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=XOKJYQNJOBDGS1WJQLVWK0KY2MX2LSBS2RQHYZ1IERYGZQM2&client_secret=A20AUTOPOO4M4NLEVPQ4L5LLCSTC4EQL2R5AOUQWQPSN0IJO&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

Send the GET request and examine the results

In [34]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d9a9831342adf00385a2883'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Malvern',
  'headerFullLocation': 'Malvern, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.8111863045, 'lng': -79.18812958073042},
   'sw': {'lat': 43.80218629549999, 'lng': -79.2005772192696}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6b9446edc76b0d771311c',
       'name': "Wendy's",
       'location': {'crossStreet': 'Morningside & Sheppard',
        'lat': 43.80744841934756,
        'lng': -79.19905558052072,
        'labeledLatLngs': [{'label': 'display'

#### Clean the json and structure it into a *pandas* dataframe.

In [35]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON
nearby_venues

reasons.count                                      reasons.items  \
0              0  [{'summary': 'This spot is popular', 'type': '...   
1              0  [{'summary': 'This spot is popular', 'type': '...   

                       referralId  \
0  e-0-4bb6b9446edc76b0d771311c-0   
1  e-0-5539e7d2498edaf4b02673ca-1   

                                    venue.categories  \
0  [{'id': '4bf58dd8d48988d16e941735', 'name': 'F...   
1  [{'id': '52f2ab2ebcbc57f1066b8b28', 'name': 'P...   

                   venue.id   venue.location.address venue.location.cc  \
0  4bb6b9446edc76b0d771311c                      NaN                CA   
1  5539e7d2498edaf4b02673ca  1315 Morningside Avenue                CA   

  venue.location.city venue.location.country venue.location.crossStreet  \
0             Toronto                 Canada     Morningside & Sheppard   
1         Scarborough                 Canada                        NaN   

   venue.location.distance                    venue.location.formattedAddress  \
0                      387                               [Toronto ON, Canada]   
1                      498  [1315 Morningside Avenue, Scarborough ON M1B 3...   

                       venue.location.labeledLatLngs  venue.location.lat  \
0  [{'label': 'display', 'lat': 43.80744841934756...           43.807448   
1  [{'label': 'display', 'lat': 43.8056297, 'lng'...           43.805630   

   venue.location.lng venue.location.postalCode venue.location.state  \
0          -79.199056                       NaN                   ON   
1          -79.200378                   M1B 3C5                   ON   

              venue.name  venue.photos.count venue.photos.groups  
0                Wendy's                   0                  []  
1  Interprovincial Group                   0                  []

In [36]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [37]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0                Wendy's  Fast Food Restaurant  43.807448 -79.199056
1  Interprovincial Group            Print Shop  43.805630 -79.200378

In [38]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

2 venues were returned by Foursquare.


#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [39]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
        

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Run above function on each neighborhood and create new dataframe called *toronto_venues*.

In [40]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )



Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

#### Let's check the size of the resulting dataframe

In [41]:
print(toronto_venues.shape)
toronto_venues.head()

(2252, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1                          Rouge, Malvern              43.806686   
2  Highland Creek, Rouge Hill, Port Union              43.784535   
3  Highland Creek, Rouge Hill, Port Union              43.784535   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.194353            Interprovincial Group       43.805630   
2              -79.160497            Royal Canadian Legion       43.782533   
3              -79.160497   Scarborough Historical Society       43.788755   
4              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.200378            Print Shop  
2       -79.163085                   Bar  
3       -79.162438        History Museum  
4       -79.189914           Pizza Place

Let's check how many venues were returned for each neighborhood

In [42]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                      9   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Downsview North, Wilson Heights                        19   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            57   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      8   
Brockton, Exhibition Place, Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      17   
CFB Toronto, Downsview East                                             2   
CN Tower, Bathurst Quay, Island airport, Harbou...                     16   
Cabbagetown, St. James Town                                            45   
Caledonia-Fairbanks                                                     5   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               7   
Central Bay Street                                                     85   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   86   
Clairlea, Golden Mile, Oakridge                                         8   
Clarks Corners, Sullivan, Tam O'Shanter                                13   
Cliffcrest, Cliffside, Scarborough Village West                         3   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     15   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                          4   
Design Exchange, Toronto Dominion Centre                              100   
Don Mills North                                                         4   
Dorset Park, Scarborough Town Centre, Wexford H...                      7   
Dovercourt Village, Dufferin                                           15   
Downsview Central                                                       4   
Downsview Northwest                                                     4   
Downsview West                                                          6   
Downsview, North Park, Upwood Park                                      4   
East Birchmount Park, Ionview, Kennedy Park                             5   
East Toronto                                                            4   
Emery, Humberlea                                                        2   
Fairview, Henry Farm, Oriole                                           66   
First Canadian Place, Underground city                                100   
Flemingdon Park, Don Mills South                                       21   
Forest Hill North, Forest Hill West                                     5   
Glencairn                                                               4   
Guildwood, Morningside, West Hill                                       7   
Harbord, University of Toronto                                         36   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                       

#### Let's find out how many unique categories can be curated from all the returned venues

In [43]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 275 uniques categories.


<a id='item4'></a>

## 4. Analyze Each Neighborhood

In [44]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.head()

Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                  0                  0        0                   0   
1                  0                  0        0                   0   
2                  0                  0        0                   0   
3                  0                  0        0                   0   
4                  0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    1   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0          0       0              0          0         0   
1          0       0              0          0         0   
2          0       0              0          0         0   
3          0       0              0          0         0   
4          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0              0         0   
1     

In [45]:
# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 
toronto_onehot.head()

Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                  0                  0        0                   0   
1                  0                  0        0                   0   
2                  0                  0        0                   0   
3                  0                  0        0                   0   
4                  0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    1   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0          0       0              0          0         0   
1          0       0              0          0         0   
2          0       0              0          0         0   
3          0       0              0          0         0   
4          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0              0         0   
1     

In [46]:
# move neighborhood column to the first column
Neighborhood = toronto_onehot['Neighborhood']

toronto_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
toronto_onehot.insert(0, 'Neighborhood', Neighborhood)

toronto_onehot.head()

Neighborhood  Accessories Store  \
0                          Rouge, Malvern                  0   
1                          Rouge, Malvern                  0   
2  Highland Creek, Rouge Hill, Port Union                  0   
3  Highland Creek, Rouge Hill, Port Union                  0   
4       Guildwood, Morningside, West Hill                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    1               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0     

And let's examine the new dataframe size.

In [47]:
toronto_onehot.shape

(2252, 275)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [48]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head(3)

Neighborhood  Accessories Store  \
0                           Adelaide, King, Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North, L'Amoreaux East, Milliken, St...                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                 0.04   
1             0.0              0.0               0.0                 0.00   
2             0.0              0.0               0.0                 0.00   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0         0.01        0.01                  0.0   
1           0.0       0.0         0.00        0.00                  0.0   
2           0.0       0.0         0.00        0.00                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0              0.03                 0.0          0.0            0.0   
1              0.00                 0.0          0.0            0.0   
2              0.00                 0.0          0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank   Bar  Baseball Field  \
0        0.0         0.0         0.0    0.02   0.0  0.04             0.0   
1        0.0         0.0         0.0    0.00   0.0  0.00             0.0   
2        0.0         0.0         0.0    0.00   0.0  0.00             0.0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0               0.0               0.0                 0.0    0.0   
1               0.0               0.0                 0.0    0.0   
2               0.0               0.0                 0.0    0.0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0              0.0       0.0         0.0                 0.0        0.0   
1              0.0       0.0         0.0                 0.0        0.0   
2              0.0       0.0         0.0                 0.0        0.0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0     0.0            0.0       0.01       0.0                  0.01   
1     0.0            0.0       0.00       0.0                  0.00   
2     0.0            0.0       0.00       0.0                  0.00   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0            0.01      0.0          0.0              0.0      0.01   
1            0.20      0.0          0.0              0.0      0.00   
2            0.00      0.0          0.0              0.0      0.00   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  Butcher  \
0          0.03           0.01       0.0          0.0       0.0      0.0   
1          0.00           0.00       0.0          0.0       0.0      0.0   
2          0.00           0.00       0.0          0.0       0.0      0.0   

   Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0  0.05                        0.0           0.0          0.0   
1  0.00                        0.0           0.0          0.0   
2  0.00                        0.0           0.0          0.0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                   0.0          0.0                 0.0             0.0   
1                   0.0          0.0                 0.2             0.0   
2                   0.0          0.0                 0.0             0.0   

   Church  Climbing Gym  Clothing Store  Cocktail Bar  Coffee Shop  \
0     0.0           0.0            0.01           0.0         0.08   
1     0.0           0.0            0.00           0.0         0.00   
2     0.0           0.0            0.00          

#### Let's confirm the new size

In [49]:
toronto_grouped.shape

(97, 275)

#### Let's print each neighborhood along with the top 5 most common venues

In [50]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.08
1             Café  0.05
2       Steakhouse  0.04
3  Thai Restaurant  0.04
4              Bar  0.04


----Agincourt----
                venue  freq
0      Sandwich Place   0.2
1      Breakfast Spot   0.2
2              Lounge   0.2
3  Chinese Restaurant   0.2
4        Skating Rink   0.2


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                             venue  freq
0                       Playground   0.5
1                             Park   0.5
2                    Metro Station   0.0
3  Molecular Gastronomy Restaurant   0.0
4       Modern European Restaurant   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                 venue  freq
0        Grocery Store  0.22
1             Pharmacy  0.11
2  Fried Chicken Joint  0.11
3       Sandwich Place  0.11
4          Video Store  0.11


----Alderwood, Long 

           venue  freq
0  Grocery Store  0.33
1           Bank  0.17
2           Park  0.17
3          Hotel  0.17
4  Shopping Mall  0.17


----Downsview, North Park, Upwood Park----
                        venue  freq
0                      Bakery  0.25
1  Construction & Landscaping  0.25
2                        Park  0.25
3            Basketball Court  0.25
4           Accessories Store  0.00


----East Birchmount Park, Ionview, Kennedy Park----
               venue  freq
0     Discount Store   0.4
1        Bus Station   0.2
2   Department Store   0.2
3        Coffee Shop   0.2
4  Mobile Phone Shop   0.0


----East Toronto----
                 venue  freq
0                 Park  0.50
1        Metro Station  0.25
2    Convenience Store  0.25
3   Mexican Restaurant  0.00
4  Monument / Landmark  0.00


----Emery, Humberlea----
                             venue  freq
0                      Fabric Shop   0.5
1                   Baseball Field   0.5
2                Accessories Store   0

                venue  freq
0   Fish & Chips Shop  0.06
1  Italian Restaurant  0.06
2                Park  0.06
3                 Pub  0.06
4         Coffee Shop  0.06


----The Danforth West, Riverdale----
                    venue  freq
0        Greek Restaurant  0.19
1             Coffee Shop  0.10
2      Italian Restaurant  0.07
3          Ice Cream Shop  0.07
4  Furniture / Home Store  0.05


----The Junction North, Runnymede----
                  venue  freq
0  Caribbean Restaurant  0.33
1               Brewery  0.33
2              Bus Line  0.33
3     Accessories Store  0.00
4     Mobile Phone Shop  0.00


----The Kingsway, Montgomery Road, Old Mill North----
               venue  freq
0              River  0.25
1               Park  0.25
2               Pool  0.25
3         Smoke Shop  0.25
4  Accessories Store  0.00


----Thorncliffe Park----
               venue  freq
0  Indian Restaurant  0.11
1     Sandwich Place  0.11
2       Burger Joint  0.11
3        Yoga Studio  0.06
4

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [51]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [52]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))    

In [53]:
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

In [54]:
for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North, L'Amoreaux East, Milliken, St...            Playground   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant                   Bar   
1          Skating Rink        Breakfast Spot        Sandwich Place   
2                  Park           Yoga Studio   Dumpling Restaurant   
3            Beer Store   Fried Chicken Joint  Fast Food Restaurant   
4                   Gym          Skating Rink              Pharmacy   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant            Steakhouse                 Hotel   
1    Chinese Restaurant             Drugstore        Discount Store   
2                 Diner        Discount Store               Dog Run   
3           Pizza Place        Sandwich Place              Pharmacy   
4           Coffee Shop    Athletics & Sports                   Pub   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant      Asian Restaurant           Burger Joint  
1               Dog Run      Doner Restaurant             Donut Shop  
2      Doner Restaurant            Donut Shop              Drugstore  
3           Video Store        Airport Lounge                  Diner  
4                  Pool        Sandwich Place                Dog Run

<a id='item5'></a>

## 5. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [55]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 2, 1, 1, 1, 0, 1, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [56]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [57]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               0                           Adelaide, King, Richmond   
1               1                                          Agincourt   
2               2  Agincourt North, L'Amoreaux East, Milliken, St...   
3               1  Albion Gardens, Beaumond Heights, Humbergate, ...   
4               1                             Alderwood, Long Branch   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                  Café       Thai Restaurant   
1                Lounge          Skating Rink        Breakfast Spot   
2            Playground                  Park           Yoga Studio   
3         Grocery Store            Beer Store   Fried Chicken Joint   
4           Pizza Place                   Gym          Skating Rink   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                   Bar   American Restaurant            Steakhouse   
1        Sandwich Place    Chinese Restaurant             Drugstore   
2   Dumpling Restaurant                 Diner        Discount Store   
3  Fast Food Restaurant           Pizza Place        Sandwich Place   
4              Pharmacy           Coffee Shop    Athletics & Sports   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                 Hotel            Restaurant      Asian Restaurant   
1        Discount Store               Dog Run      Doner Restaurant   
2               Dog Run      Doner Restaurant            Donut Shop   
3              Pharmacy           Video Store        Airport Lounge   
4                   Pub                  Pool        Sandwich Place   

  10th Most Common Venue  
0           Burger Joint  
1             Donut Shop  
2              Drugstore  
3                  Diner  
4                Dog Run

In [58]:
# merge neighborhoods_venues_sorted with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = neighborhoods_venues_sorted.join(toronto_data.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(2)

Cluster Labels              Neighborhood 1st Most Common Venue  \
0               0  Adelaide, King, Richmond           Coffee Shop   
1               1                 Agincourt                Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant                   Bar   
1          Skating Rink        Breakfast Spot        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant            Steakhouse                 Hotel   
1    Chinese Restaurant             Drugstore        Discount Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
0            Restaurant      Asian Restaurant           Burger Joint   
1               Dog Run      Doner Restaurant             Donut Shop   

  PostalCode           Borough   Latitude  Longitude  
0        M5H  Downtown Toronto  43.650571 -79.384568  
1        M1S       Scarborough  43.794200 -79.262029

Finally, let's visualize the resulting clusters

In [59]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item6'></a>

## 6. Examine Clusters

Examine each cluster and determine the discriminating venue categories that distinguish each cluster.

#### Cluster 1

In [60]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[1:12]]

Neighborhood  \
0                            Adelaide, King, Richmond   
6                                     Bayview Village   
8                                         Berczy Park   
9                         Birch Cliff, Cliffside West   
11       Brockton, Exhibition Place, Parkdale Village   
12  Business Reply Mail Processing Centre 969 Eastern   
14  CN Tower, Bathurst Quay, Island airport, Harbo...   
15                        Cabbagetown, St. James Town   
18                                          Cedarbrae   
19                                 Central Bay Street   
20          Chinatown, Grange Park, Kensington Market   
21                                           Christie   
22                               Church and Wellesley   
23                    Clairlea, Golden Mile, Oakridge   
25    Cliffcrest, Cliffside, Scarborough Village West   
26                     Commerce Court, Victoria Hotel   
28                                   Davisville North   
31           Design Exchange, Toronto Dominion Centre   
32                                    Don Mills North   
33  Dorset Park, Scarborough Town Centre, Wexford ...   
34                       Dovercourt Village, Dufferin   
35                                  Downsview Central   
38                 Downsview, North Park, Upwood Park   
41                                   Emery, Humberlea   
42                       Fairview, Henry Farm, Oriole   
43             First Canadian Place, Underground city   
45                Forest Hill North, Forest Hill West   
46                                          Glencairn   
48                     Harbord, University of Toronto   
49  Harbourfront East, Toronto Islands, Union Station   
50                          Harbourfront, Regent Park   
51                      High Park, The Junction South   
52             Highland Creek, Rouge Hill, Port Union   
53                                  Hillcrest Village   
54       Humber Bay Shores, Mimico South, New Toronto   
55  Humber Bay, King's Mill Park, Kingsway Park So...   
57                                 Humewood-Cedarvale   
61                   Lawrence Heights, Lawrence Manor   
64                           Little Portugal, Trinity   
65                                  Maryvale, Wexford   
67                                 North Toronto West   
68                                          Northwest   
69                    Northwood Park, York University   
70                             Parkdale, Roncesvalles   
76                                 Runnymede, Swansea   
77                           Ryerson, Garden District   
78                                Scarborough Village   
79                                     St. James Town   
80                    Stn A PO Boxes 25 The Esplanade   
81                                    Studio District   
83                                        The Beaches   
85                       The Danforth West, Riverdale   
86                      The Junction North, Runnymede   
89                                   Victoria Village   
95                                   Woodbine Heights   

         1st Most Common Venue          2nd Most Common Venue  \
0                  Coffee Shop                           Café   
6                         Café                           Bank   
8                  Coffee Shop                   Cocktail Bar   
9              College Stadium          General Entertainment   
11              Breakfast Spot                           Café   
12          Light Rail Station                 Farmers Market   
14             Airport Service               Airport Terminal   
15                  Restaurant                           Café   
18            Hakka Restaurant            Fried Chicken Joint   
19                 Coffee Shop                           Café   
20                        Café  Vegetarian / Vegan Restaurant   
21               Grocery Store                           Café   
22 

#### Cluster 2

In [61]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[1:12]]

Neighborhood 1st Most Common Venue  \
1                                           Agincourt                Lounge   
3   Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                              Alderwood, Long Branch           Pizza Place   
5     Bathurst Manor, Downsview North, Wilson Heights           Coffee Shop   
7                   Bedford Park, Lawrence Manor East    Italian Restaurant   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...              Pharmacy   
17              Canada Post Gateway Processing Centre                 Hotel   
24            Clarks Corners, Sullivan, Tam O'Shanter              Pharmacy   
27                                         Davisville           Pizza Place   
29  Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
30     Del Ray, Keelesdale, Mount Dennis, Silverthorn            Restaurant   
36                                Downsview Northwest  Gym / Fitness Center   
37                                     Downsview West         Grocery Store   
39        East Birchmount Park, Ionview, Kennedy Park        Discount Store   
44                   Flemingdon Park, Don Mills South                   Gym   
47                  Guildwood, Morningside, West Hill    Mexican Restaurant   
56                                      Humber Summit           Pizza Place   
59  Kingsway Park South West, Mimico NW, The Queen...                   Gym   
60                                    L'Amoreaux West  Fast Food Restaurant   
63                                            Leaside           Coffee Shop   
72                                       Queen's Park           Coffee Shop   
82                The Annex, North Midtown, Yorkville                  Café   
84                     The Beaches West, India Bazaar          Liquor Store   
88                                   Thorncliffe Park     Indian Restaurant   
90                                          Westmount           Pizza Place   
91                                   Willowdale South           Coffee Shop   
92                                    Willowdale West              Pharmacy   
93                                             Woburn           Coffee Shop   
94                    Woodbine Gardens, Parkview Hill  Fast Food Restaurant   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
1           Skating Rink        Breakfast Spot               Sandwich Place   
3             Beer Store   Fried Chicken Joint         Fast Food Restaurant   
4                    Gym          Skating Rink                     Pharmacy   
5          Shopping Mall        Sandwich Place                  Supermarket   
7            Coffee Shop          Liquor Store                  Pizza Place   
10            Beer Store           Pizza Place            Convenience Store   
17           Coffee Shop  Gym / Fitness Center               Sandwich Place   
24           Pizza Place                  Bank              Thai Restaurant   
27        Sandwich Place          Dessert Shop             Sushi Restaurant   
29                   Pub   American Restaurant                   Bagel Shop   
30        Sandwich Place           Coffee Shop                          Bar   
36         Grocery Store          Liquor Store               Discount Store   
37                  Park                  Bank                        Hotel   
39      Department Store           Bus Station                  Coffee Shop   
44      Asian Restaurant           Coffee Shop                   Beer Store   
47           Pizza Place   Rental Car Location               Breakfast Spot   
56   Empanada Restaurant   Dumpling Restaurant                        Diner   
59     Convenience Store       Supplement Shop                Burrito Place   
60    Chinese Restaurant              Pharmacy       Thrift / Vintage Store   
63   Sporting Goods Shop          Burger Joint       Furniture / Home Store   
72         

#### Cluster 3

In [62]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[1:12]]

Neighborhood 1st Most Common Venue  \
2   Agincourt North, L'Amoreaux East, Milliken, St...            Playground   
13                        CFB Toronto, Downsview East               Airport   
16                                Caledonia-Fairbanks                  Park   
40                                       East Toronto                  Park   
58  Kingsview Village, Martin Grove Gardens, Richv...           Pizza Place   
62                                      Lawrence Park                  Park   
66                        Moore Park, Summerhill East            Playground   
71                                          Parkwoods                  Park   
73                                           Rosedale                  Park   
87      The Kingsway, Montgomery Road, Old Mill North                 River   
96                                    York Mills West                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
2                   Park           Yoga Studio          Dumpling Restaurant   
13                  Park           Yoga Studio  Eastern European Restaurant   
16         Women's Store                Market         Fast Food Restaurant   
40     Convenience Store         Metro Station                  Yoga Studio   
58     Mobile Phone Shop                  Park                    Drugstore   
62           Swim School              Bus Line                  Yoga Studio   
66                  Park           Yoga Studio          Dumpling Restaurant   
71                  Pool     Food & Drink Shop                  Yoga Studio   
73            Playground              Building                        Trail   
87                  Park            Smoke Shop                         Pool   
96                  Bank     Convenience Store                  Yoga Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2                         Diner        Discount Store               Dog Run   
13               Discount Store               Dog Run      Doner Restaurant   
16                 Gourmet Shop           Fabric Shop           Event Space   
40          Dumpling Restaurant        Discount Store               Dog Run   
58                        Diner        Discount Store               Dog Run   
62                    Drugstore        Discount Store               Dog Run   
66                        Diner        Discount Store               Dog Run   
71          Dumpling Restaurant        Discount Store               Dog Run   
73                  Yoga Studio             Drugstore                 Diner   
87                   Donut Shop          Dessert Shop    Dim Sum Restaurant   
96  Eastern European Restaurant               Dog Run      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2       Doner Restaurant            Donut Shop              Drugstore  
13            Donut Shop             Drugstore    Dumpling Restaurant  
16  Ethiopian Restaurant   Empanada Restaurant      Electronics Store  
40      Doner Restaurant            Donut Shop              Drugstore  
58      Doner Restaurant            Donut Shop    Dumpling Restaurant  
62      Doner Restaurant            Donut Shop    Dumpling Restaurant  
66      Doner Restaurant            Donut Shop              Drugstore  
71      Doner Restaurant            Donut Shop              Drugstore  
73        Discount Store               Dog Run       Doner Restaurant  
87                 Diner        Discount Store                Dog Run  
96            Donut Shop             Drugstore    Dumpling Restaurant

#### Cluster 4

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[1:12]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
75  Rouge, Malvern  Fast Food Restaurant            Print Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
75                 Diner        Discount Store               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
75      Doner Restaurant            Donut Shop             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
75   Dumpling Restaurant           Dessert Shop

#### Cluster 5

In [64]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[1:12]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
74     Roselawn                Garden           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
74   Dumpling Restaurant        Discount Store               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
74      Doner Restaurant            Donut Shop             Drugstore   

          9th Most Common Venue 10th Most Common Venue  
74  Eastern European Restaurant     Dim Sum Restaurant

### Thank You

* References: Notebook created by [Alex Aklson](https://www.linkedin.com/in/aklson/) and [Polong Lin](https://www.linkedin.com/in/polonglin/) for the course on **Coursera** called *Applied Data Science Capstone*

<hr>